File: experiment_0/cpp/computational_consciousness.cpp
#include <SFML/Graphics.hpp>
#include <Eigen/Dense>
#include <tiny_dnn/tiny_dnn.h>
#include <thread>
#include <mutex>
#include <queue>
#include <vector>
#include <iostream>
#include <random>
#include <functional>
#include <atomic>
#include <condition_variable>
#include <cmath>
#include <algorithm>
#include <chrono>

using namespace std;
using namespace sf;
using namespace Eigen;
using namespace tiny_dnn;
using namespace tiny_dnn::layers;
using namespace std::chrono;

constexpr int SCREEN_WIDTH = 800;
constexpr int SCREEN_HEIGHT = 600;
const sf::Color BACKGROUND_COLOR = sf::Color(255, 255, 255);
const sf::Color AGENT_COLOR = sf::Color(0, 0, 255);
const sf::Color FOOD_COLOR = sf::Color(0, 255, 0);
constexpr int AGENT_SIZE = 10;
constexpr int FOOD_SIZE = 5;
constexpr int MATURITY_TIME = 10; // in seconds
constexpr int MAX_AGE = 180; // in seconds
constexpr int NUM_FOOD_ITEMS = 1000; // Number of food items to add

mutex mtx;
condition_variable cv;
priority_queue<pair<int, function<void()>>> taskQueue;
mutex agentsMutex;

struct Food {
    sf::Vector2f position;

    Food(float x, float y) : position(x, y) {}

    void draw(RenderWindow& window) {
        CircleShape shape(FOOD_SIZE);
        shape.setFillColor(FOOD_COLOR);
        shape.setPosition(position.x, position.y);
        window.draw(shape);
        // cout << "Drawing food at position: " << position.x << ", " << position.y << endl; // Comment out for less verbosity
    }
};

struct Task {
    string name;
    int priority;
    double complexity;
    function<void()> action;

    Task(string n, int p, double c, function<void()> a) : name(move(n)), priority(p), complexity(c), action(move(a)) {}

    bool operator<(const Task& other) const {
        return priority > other.priority; // Higher priority first
    }
};

Vector4f getRealTimeData();
float predictPriority(network<sequential>& net, const Vector4f& data);

class Agent {
public:
    string name;
    sf::Vector2f position;
    sf::Vector2f direction;
    priority_queue<Task> tasks;
    atomic<float> energy;
    bool alive = true;
    float stress = 0;
    time_point<steady_clock> birthTime;
    time_point<steady_clock> lastReproduceTime;

    Agent(string n, float x, float y, float e) : name(std::move(n)), position(x, y), direction(1, 0), energy(e) {
        birthTime = steady_clock::now();
        lastReproduceTime = birthTime;
        cout << "Agent " << name << " created at position (" << x << ", " << y << ") with energy " << e << endl; // Debug statement
    }


    Agent(Agent&& other) noexcept 
        : name(std::move(other.name)), position(other.position), direction(other.direction), tasks(std::move(other.tasks)), 
          energy(other.energy.load()), alive(other.alive), stress(other.stress), birthTime(other.birthTime), lastReproduceTime(other.lastReproduceTime) {}

    Agent& operator=(Agent&& other) noexcept {
        if (this != &other) {
            name = std::move(other.name);
            position = other.position;
            direction = other.direction;
            tasks = std::move(other.tasks);
            energy.store(other.energy.load());
            alive = other.alive;
            stress = other.stress;
            birthTime = other.birthTime;
            lastReproduceTime = other.lastReproduceTime;
        }
        return *this;
    }

    void addTask(const Task& task) {
        lock_guard<mutex> lock(mtx);
        tasks.push(task);
        cv.notify_all();
        cout << "Task added: " << task.name << " with priority " << task.priority << endl;
    }

    void processTasks() {
        while (alive) {
            unique_lock<mutex> lock(mtx);
            if (tasks.empty()) {
                cout << "No tasks to process for agent: " << name << endl;
                return;
            }

            Task task = tasks.top();
            tasks.pop();
            cout << "Popped task: " << task.name << " with priority " << task.priority << endl;
            lock.unlock();
            if (energy >= task.complexity) {
                cout << "Executing task: " << task.name << " with priority " << task.priority << endl;
                try {
                    task.action();
                } catch (const std::exception& e) {
                    cout << "Exception in task action: " << e.what() << endl;
                } catch (...) {
                    cout << "Unknown exception in task action" << endl;
                }
                float currentEnergy = energy.load();
                energy.store(currentEnergy - task.complexity);
            } else {
                cout << "Not enough energy for task: " << task.name << endl;
            }
        }
    }

    void simulateLife() {
        thread taskProcessor(&Agent::processTasks, this);
        taskProcessor.detach();
    }

    void draw(RenderWindow& window) {
        CircleShape shape(AGENT_SIZE);
        shape.setFillColor(AGENT_COLOR);
        shape.setPosition(position.x, position.y);
        window.draw(shape);
        // cout << "Drawing agent " << name << " at position: " << position.x << ", " << position.y << endl; // Comment out for less verbosity
    }

    void moveAgent() {
        cout << "moveAgent called" << endl; // Debug statement
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<> dist(-1, 1);
        direction = sf::Vector2f(dist(gen), dist(gen));
        position += direction;
        position = sf::Vector2f(max(0.f, min(position.x, float(SCREEN_WIDTH))),
                            max(0.f, min(position.y, float(SCREEN_HEIGHT))));
        cout << name << " moved to " << position.x << ", " << position.y << endl;
    }

    void stayPut() {
        cout << "stayPut called" << endl; // Debug statement
        cout << name << " is staying put." << endl;
    }

    void reproduce(vector<Agent>& agents) {
        cout << "reproduce called" << endl; // Debug statement
        auto now = steady_clock::now();
        auto age = chrono::duration_cast<chrono::seconds>(now - birthTime).count();
        auto timeSinceLastReproduce = chrono::duration_cast<chrono::seconds>(now - lastReproduceTime).count();

        cout << "Agent age: " << age << " seconds" << endl;
        cout << "Time since last reproduce: " << timeSinceLastReproduce << " seconds" << endl;
        cout << "Agent energy: " << energy.load() << endl;

        if (age >= MATURITY_TIME && timeSinceLastReproduce >= MATURITY_TIME && energy > 50) {
            cout << "Conditions met for reproduction" << endl;

            string childName = name + "_child";
            float childX = position.x;
            float childY = position.y;
            float childEnergy = 50;

            cout << "Creating new agent with name: " << childName << ", position: (" << childX << ", " << childY << "), energy: " << childEnergy << endl;

            try {
                agents.push_back(Agent(childName, childX, childY, childEnergy));
                cout << "New agent added" << endl;
            } catch (const std::exception& e) {
                cout << "Exception in reproduce: " << e.what() << endl;
            } catch (...) {
                cout << "Unknown exception in reproduce" << endl;
            }

            float currentEnergy = energy.load();
            energy.store(currentEnergy - 50);
            lastReproduceTime = now;
            cout << name << " has reproduced." << endl;
        } else {
            if (age < MATURITY_TIME) {
                cout << "Agent not mature enough to reproduce." << endl;
            }
            if (timeSinceLastReproduce < MATURITY_TIME) {
                cout << "Not enough time since last reproduce." << endl;
            }
            if (energy <= 50) {
                cout << "Not enough energy to reproduce." << endl;
            }
        }
    }

    void senseEnvironment(const vector<Food>& foods, const vector<Agent>& agents) {
        vision(foods);
        hearing(agents);
        touch(agents);
    }

    void vision(const vector<Food>& foods) {
        cout << "vision called" << endl; // Debug statement
        float visionRange = 50;
        float visionAngle = M_PI / 4;
        for (const auto& food : foods) {
            sf::Vector2f toFood = food.position - position;
            float distance = sqrt(toFood.x * toFood.x + toFood.y * toFood.y);
            if (distance < visionRange) {
                float angle = acos((direction.x * toFood.x + direction.y * toFood.y) /
                                   (sqrt(direction.x * direction.x + direction.y * direction.y) * distance));
                if (angle < visionAngle) {
                    float currentEnergy = energy.load();
                    energy.store(currentEnergy + 10);
                    cout << name << " found food by vision and increased energy to " << energy << endl;
                    break;
                }
            }
        }
    }

    void hearing(const vector<Agent>& agents) {
        cout << "hearing called" << endl; // Debug statement
        float hearingRange = 100;
        for (const auto& agent : agents) {
            if (this != &agent) {
                float distance = sqrt((position.x - agent.position.x) * (position.x - agent.position.x) +
                                      (position.y - agent.position.y) * (position.y - agent.position.y));
                if (distance < hearingRange) {
                    cout << name << " hears " << agent.name << endl;
                }
            }
        }
    }

    void touch(const vector<Agent>& agents) {
        cout << "touch called" << endl; // Debug statement
        for (const auto& agent : agents) {
            if (this != &agent) {
                float distance = sqrt((position.x - agent.position.x) * (position.x - agent.position.x) +
                                      (position.y - agent.position.y) * (position.y - agent.position.y));
                if (distance < AGENT_SIZE) {
                    cout << name << " touches " << agent.name << endl;
                    stress += 10;
                }
            }
        }
    }

    void decideNextAction(vector<Agent>& agents, network<sequential>& net) {
        cout << "decideNextAction called" << endl; // Debug statement
        random_device rd;
        mt19937 gen(rd());
        uniform_real_distribution<> dis(0.1, 0.5);

        vector<function<void()>> actions = {bind(&Agent::moveAgent, this), bind(&Agent::stayPut, this)};
        auto now = steady_clock::now();
        auto age = chrono::duration_cast<chrono::seconds>(now - birthTime).count();

        if (age >= MATURITY_TIME && energy > 50) {
            actions.push_back(bind(&Agent::reproduce, this, ref(agents)));
        }

        for (const auto& action : actions) {
            float priority = predictPriority(net, getRealTimeData()) - float(stress);
            addTask(Task("Action", priority, dis(gen), action));
        }

        if (age >= MAX_AGE) {
            alive = false;
            cout << name << " has reached maximum age and died." << endl;
        }
    }

    void update(const vector<Food>& foods, vector<Agent>& agents, network<sequential>& net) {
        if (alive) {
            cout << "Updating agent: " << name << endl;
            senseEnvironment(foods, agents);
            decideNextAction(agents, net);
            cout << "Processing tasks for agent: " << name << endl;
            processTasks();
        }
    }
};

network<sequential> setupNeuralNetwork() {
    network<sequential> net;
    net << fully_connected_layer(4, 20) << tanh_layer()
        << fully_connected_layer(20, 1) << sigmoid_layer();
    return net;
}

Vector4f getRealTimeData() {
    random_device rd;
    mt19937 gen(rd());
    normal_distribution<> d_light(500, 50), d_sound(300, 50), d_touch(150, 30);
    uniform_real_distribution<> stress(0, 1);
    Vector4f data(d_light(gen), d_sound(gen), d_touch(gen), stress(gen));
    return data;
}

float predictPriority(network<sequential>& net, const Vector4f& data) {
    vec_t input = {data[0], data[1], data[2], data[3]};
    return net.predict(input)[0];
}

vector<Agent> agents;
vector<Food> foods;
network<sequential> net;

void initializeSimulation() {
    net = setupNeuralNetwork();
    agents.reserve(10000); // Preallocate space to avoid reallocation
    agents.emplace_back("Agent1", 400, 300, 100);

    // Generate random food items
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> disX(0, SCREEN_WIDTH);
    uniform_int_distribution<> disY(0, SCREEN_HEIGHT);
    for (int i = 0; i < NUM_FOOD_ITEMS; ++i) {
        foods.emplace_back(disX(gen), disY(gen));
    }

    // Comment out threading for now
    // for (auto& agent : agents) {
    //     agent.simulateLife();
    // }
}

void runSimulation() {
    RenderWindow window(VideoMode(SCREEN_WIDTH, SCREEN_HEIGHT), "Agent-Based Simulation");
    window.setFramerateLimit(60);

    initializeSimulation();
    cout << "Simulation initialized" << endl;

    while (window.isOpen()) {
        Event event;
        while (window.pollEvent(event)) {
            if (event.type == Event::Closed)
                window.close();
        }

        window.clear(BACKGROUND_COLOR);

        for (auto& food : foods) {
            food.draw(window);
        }

        {
            lock_guard<mutex> lock(agentsMutex); // Lock the agents vector during the draw call
            for (auto& agent : agents) {
                cout << "Updating and drawing agent: " << agent.name << endl; // Debug statement
                agent.update(foods, agents, net);
                agent.draw(window);
            }
        }

        window.display();
        cout << "Window updated and displayed" << endl;
    }
}

int main() {
    cout << "Starting simulation" << endl;
    runSimulation();
    return 0;
}

File: experiment_0/python/computational_consciousness.py
import threading
import queue
import random
import time
import numpy as np
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Dense, Input, Concatenate
from tensorflow.keras.callbacks import ModelCheckpoint
import pygame

# Initialize Pygame
pygame.init()

# Set display dimensions
SCREEN_WIDTH, SCREEN_HEIGHT = 800, 600
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
pygame.display.set_caption("Agent-Based Simulation with MTPQ")

# Colors and other constants
BACKGROUND_COLOR = (255, 255, 255)
AGENT_COLOR = (0, 0, 255)
FOOD_COLOR = (0, 255, 0)
AGENT_SIZE = 10
FOOD_SIZE = 5

memory = {}
sensory_data = {'light': 450, 'sound': 300, 'touch': 150}  # Expanded sensory input data

class Task:
    """A task class with name, priority, complexity, and influenced by environmental data."""
    def __init__(self, name, priority, complexity, action=None, agent=None):
        self.name = name
        self.priority = priority
        self.complexity = complexity
        self.action = action
        self.agent = agent

    def process(self):
        if self.agent and self.agent.energy > self.complexity:
            if self.action:
                self.action()
            self.agent.energy -= self.complexity
            print(f"Processed {self.name} with priority {self.priority} by {self.agent.name}")
        else:
            print(f"Insufficient energy to process {self.name}")
        return self.agent and self.agent.energy > 0

    def __lt__(self, other):
        """Define a less-than method to compare tasks based on priority."""
        return self.priority < other.priority

class Agent:
    def __init__(self, name, x, y, start_energy=100):
        self.name = name
        self.position = np.array([x, y])
        self.task_queue = queue.PriorityQueue()
        self.energy = start_energy
        self.memory = []
        self.alive = True

    def add_task(self, task):
        self.task_queue.put(task)

    def process_tasks(self):
        while not self.task_queue.empty() and self.energy > 0:
            _, task = self.task_queue.get()
            task.process()

    def update(self):
        if self.alive:
            self.sense_environment()
            self.decide_next_action()
            self.process_tasks()
            if self.energy <= 0:
                self.alive = False
                print(f"{self.name} has died.")

    def move(self):
        direction = np.random.randint(-1, 2, 2)
        self.position = np.clip(self.position + direction, 0, [SCREEN_WIDTH, SCREEN_HEIGHT])
        print(f"{self.name} moved to {self.position}")

    def stay_put(self):
        print(f"{self.name} is staying put.")

    def reproduce(self, agents):
        if self.energy > 50:  # Reproduction requires at least 50 energy
            new_agent = Agent(f"{self.name}_child", self.position[0], self.position[1], start_energy=50)
            agents.append(new_agent)
            self.energy -= 50
            print(f"{self.name} has reproduced.")

    def sense_environment(self):
        data = get_real_time_data()
        priority = predict_priority(data)
        complexity = random.uniform(0.1, 0.5)
        task = Task("Sense Environment", priority, complexity, self.action, self)
        self.add_task(task)

    def decide_next_action(self):
        actions = [self.move, self.stay_put]
        action_names = ["Move", "Stay Put"]
        real_time_data = get_real_time_data()

        # Add reproduction as an action based on certain conditions
        if self.energy > 50:
            actions.append(lambda: self.reproduce(all_agents))
            action_names.append("Reproduce")

        for action, action_name in zip(actions, action_names):
            priority = predict_priority(real_time_data)
            complexity = random.uniform(0.5, 1.5)
            task = Task(action_name, priority, complexity, action, self)
            self.add_task(task)

    def action(self):
        print(f"{self.name} is interacting with the environment.")

    def draw(self):
        pygame.draw.circle(screen, AGENT_COLOR, self.position, AGENT_SIZE)

class Food:
    def __init__(self, x, y):
        self.position = np.array([x, y])

    def draw(self):
        pygame.draw.circle(screen, FOOD_COLOR, self.position, FOOD_SIZE)

def create_multisensory_model():
    """Create a neural network model to predict task priority based on multiple sensory inputs."""
    visual_input = Input(shape=(1,))
    auditory_input = Input(shape=(1,))
    tactile_input = Input(shape=(1,))
    stress_input = Input(shape=(1,))

    # Enhanced pathways for different inputs
    visual_path = Dense(20, activation='relu')(visual_input)
    auditory_path = Dense(20, activation='relu')(auditory_input)
    tactile_path = Dense(20, activation='relu')(tactile_input)
    stress_path = Dense(20, activation='relu')(stress_input)

    # Concatenate all pathways and add more complex layers
    concatenated = Concatenate()([visual_path, auditory_path, tactile_path, stress_path])
    dense_layer = Dense(50, activation='relu')(concatenated)
    output = Dense(1, activation='sigmoid')(dense_layer)

    model = Model(inputs=[visual_input, auditory_input, tactile_input, stress_input], outputs=output)
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

priority_model = create_multisensory_model()

def get_real_time_data():
    """Simulate real-time sensory data."""
    light_level = np.random.normal(loc=500, scale=50)
    sound_level = np.random.normal(loc=300, scale=50)
    touch_pressure = np.random.normal(loc=150, scale=30)
    stress_level = np.random.choice(['high', 'medium', 'low'], p=[0.1, 0.3, 0.6])
    return {'light': light_level, 'sound': sound_level, 'touch': touch_pressure, 'stress': stress_level}

def predict_priority(environmental_data):
    """Predict task priority using the neural network model."""
    data = [
        np.array([[environmental_data['light']]]),
        np.array([[environmental_data['sound']]]),
        np.array([[environmental_data['touch']]]),
        np.array([[1.0 if environmental_data['stress'] == 'high' else 0.0]])
    ]
    return priority_model.predict(data)[0][0]

def add_tasks_to_queue(q, context):
    """Populate the priority queue with tasks based on the provided context and predicted priorities."""
    real_time_data = get_real_time_data()
    num_tasks = 30 if context == 'high_stress' else 15
    complexities = [random.uniform(1, 3) if context == 'high_stress' else random.uniform(0.5, 1.5) for _ in range(num_tasks)]

    for i in range(num_tasks):
        predicted_priority = predict_priority(real_time_data) * 10  # Scale the priority
        complexity = complexities[i]
        task = Task(f"Task-{i}", predicted_priority, complexity)
        q.put((task.priority, task))

def process_tasks_from_queue(q):
    """Process tasks from the queue, adjusting for memory of past tasks."""
    while not q.empty():
        try:
            _, task = q.get(timeout=1)
            if task.name in memory:
                task.complexity *= 0.9  # Reduce complexity if task is remembered
            task.process()
            memory[task.name] = 'processed'
            q.task_done()
        except queue.Empty:
            break

def evaluate_strategy(num_workers, context):
    """Evaluate a strategy based on the number of workers and context."""
    pq = queue.PriorityQueue()
    add_tasks_to_queue(pq, context)

    start_time = time.time()
    threads = []
    for _ in range(num_workers):
        thread = threading.Thread(target=process_tasks_from_queue, args=(pq,))
        thread.start()
        threads.append(thread)

    for thread in threads:
        thread.join()
    total_time = time.time() - start_time
    reflect_on_performance(total_time)
    return total_time

def genetic_algorithm(population, context, generations=10):
    """Genetic algorithm to evolve the number of worker threads."""
    for generation in range(generations):
        fitness_scores = [(evaluate_strategy(individual, context), individual) for individual in population]
        fitness_scores.sort()

        population = [individual for _, individual in fitness_scores[:max(len(population)//2, 2)]]

        if len(population) < 2:
            population += [random.randint(1, 10) for _ in range(2 - len(population))]

        next_gen = []
        while len(next_gen) < len(fitness_scores):
            if len(population) > 1:
                parent1, parent2 = random.sample(population, 2)
                child = (parent1 + parent2) // 2
                if random.random() < 0.1:
                    child += random.choice([-1, 1])
                    child = max(1, child)
                next_gen.append(child)
            else:
                child = population[0] + random.choice([-1, 1])
                child = max(1, child)
                next_gen.append(child)

        population = next_gen
        print(f"Generation {generation + 1}: Best time {fitness_scores[0][0]} with {fitness_scores[0][1]} threads")
    return population

def reflect_on_performance(total_time):
    """Adjust strategies based on performance metrics."""
    acceptable_threshold = 10  # Example threshold
    if total_time > acceptable_threshold:
        print("Performance below expectations, consider strategy adjustments.")

def run_simulation():
    global all_agents
    clock = pygame.time.Clock()
    all_agents = [Agent(f"Agent_{i}", random.randint(0, SCREEN_WIDTH), random.randint(0, SCREEN_HEIGHT)) for i in range(10)]
    food_items = [Food(random.randint(0, SCREEN_WIDTH), random.randint(0, SCREEN_HEIGHT)) for _ in range(20)]

    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False

        screen.fill(BACKGROUND_COLOR)

        for food in food_items:
            food.draw()

        alive_agents = []
        for agent in all_agents:
            agent.update()
            if agent.alive:
                agent.draw()
                alive_agents.append(agent)
        all_agents = alive_agents

        pygame.display.flip()
        clock.tick(60)

    pygame.quit()

if __name__ == '__main__':
    initial_population = [random.randint(1, 10) for _ in range(10)]
    context = 'high_stress'  # Can be changed to 'normal' or other contexts
    best_strategies = genetic_algorithm(initial_population, context)
    print("Best strategies:", best_strategies)
    run_simulation()
